### PR TITLE
Added multiple flags to moodle

### LINKED ISSUE #17: Moodle exam generation should support multiple flags in a single exam question
Currently, only one flag per question is supported when generating the Moodle exam.
However, sometimes the build process can embed many flags at once, and the assignment itself might require this.

Due to limitations of the Moodle, we need to connect multiple flags into single question (there is no other way to verify that submitted flag is connected to the provided question/output files)

Functionality for this is related here:
https://github.com/ouspg/ainigma/blob/fa4e9a470614c37ad0aa1990407fcfc431056b78/src/moodle.rs#L66

Basically, for example, if there are three different flags, they should be submitted in specific format so that also partial answer is accepted.

E.g. one from three provides 1 points out of three points and so on.. 
 * The exam should note all the orders or combinations when submitting the flag and give correct points.
 * E.g. submitting three flags in straing as `flag{value1} flag{value2} flag{value3}` would attempt to provide three different flags. If the flags are correct, the answer should be correct in any order and these should be pre-generated for the Moodle exam.

### PR DESCRIPTION
**Pull request by PootisHunter**
Closes #17 

-Added multiple flag support for answers and a test that shows functionality (test_multiple_flags)

- Current separator for flags:  ";"

- Using itertools crate to implement combinations and permutations for flags

Generating answers for 3 flags currently adds total 6+12+12 = 30 answers
- 6 answers for 33 fraction ( 3 flag_string and 3 encase_flag answers)
- 12 answers for 67 fraction ( 6 flag_string and 6 encase_flag answers)
- 12 answers for 100 fraction  ( 6 flag_string and 6 encase_flag answers)

### CODE REVIEW

File: src/moodle.rs [hunk 1]
```diff
@@ -82,3 +85,79 @@ pub fn create_exam(
...
+                    "Correct!".to_string().into(),
+                ));
+            }
+        }
+    }
+    answers
+}
+
+#[cfg(test)]
+mod tests {
```
Comment (Nicceboy) [inline src/moodle.rs:132]: I think we might need to add at least one test e.g. with 3 flags to guarantee that flags work as intended and produce the output deterministically and the final exam has exactly the amount of correct unique permutations (no duplicates, for example).

Comment (Nicceboy) [reply src/moodle.rs:132]: We also need a test when student submits 1, 2 or 3 flags and see that all are accepted if the flag is correct.

File: src/moodle.rs [hunk 2]
```diff
@@ -63,7 +65,8 @@ pub fn create_exam(
                         ),
                     ]
                 } else {
-                    todo!("Multiple flags in subtasks not supported yet")
+                    // Adds 1-inf flags as answer with chosen separator
+                    process_multiple_flags(item.flags.clone(), ";")
```
Comment (Nicceboy) [inline src/moodle.rs]: We can make separator as constant in this file and pass it as reference for this. I believe space is better separator in this context (for readability and clarity). Flags should never have spaces in them, while other characters are possible.

File: src/moodle.rs [hunk 3]
```diff
@@ -82,3 +85,79 @@ pub fn create_exam(
...
+    for r in 1..=total_flags {
+        for combination in flags.iter().combinations(r) {
+            for perm in combination.iter().permutations(r) {
+                let perm_flags: Vec<Flag> =
+                    perm.iter().cloned().map(|&flag| flag.clone()).collect();
+                let encased_combined_answer = encase_each_flag(&perm_flags, separator); // Pass as a slice
+                let combined_answer = join_flags(&perm_flags, separator); // Pass as a slice
+
+                // Calculate points based on the number of flags
+                let points = ((r as f64 / total_flags as f64) * 100.0).round() as u8;
```
Comment (Nicceboy) [inline src/moodle.rs:113]: We need to round with 4 decimals (at least that is how Moodle does it internally)

**Review by Nicceboy — CHANGES_REQUESTED** [review]

File: src/moodle.rs (hunk 1 above)
Comment (Nicceboy) [reply src/moodle.rs:132]: These tests can be mainly implemented with constructed strings

File: src/moodle.rs (hunk 3 above)
Comment (PootisHunter) [reply src/moodle.rs:113]: What this code does is give answer fraction value between 0-100 depending on how many correct flags are included

1 / 3 correct flags giving 33
2 / 3 correct flags giving 67
3 / 3 correct flags giving 100 and etc.

the Answer  type fraction values are as u8 so rounding with decimals is not required

File: src/moodle.rs (hunk 2 above)
Comment (PootisHunter) [reply src/moodle.rs]: changed separator to " "

File: src/moodle.rs (hunk 1 above)
Comment (PootisHunter) [reply src/moodle.rs:132]: changed tests to check 1, 2, 3 flag answers and make sure they contain

- appropriate fraction values depending on the amount of flags
- contained appropriate amount of flag_prefixes for the fraction value

File: src/moodle.rs (hunk 3 above)
Comment (Nicceboy) [reply src/moodle.rs:113]: Maybe we can test it against Moodle with a real exam and see if works, and address it later if it does not like the fractions on some cases.

File: src/moodle.rs [hunk 4]
```diff
@@ -82,3 +85,144 @@ pub fn create_exam(
...
+                100 => {
+                    assert!(answer.text.contains("task2:") && answer.text.contains("task3:"));
+                }
+                _ => {
+                    unreachable!("Unexpected fraction value encountered in test")
+                }
+            }
+        }
+    }
+}
```
Comment (Nicceboy) [inline src/moodle.rs:228]: OK. This does not ensure that there are no duplicates but maybe we can trust the logic of itertools. However, tests should always ensure that if we made changes in the code, the end functionality remains and we get confidence that system still works, so tests should be strict enough.

**Review by Nicceboy — APPROVED** [review]

**Pull request merged**